src/libclipper/src/containers.cpp

```diff
@@ -9,15 +10,70 @@
 #include <clipper/constants.hpp>
 #include <clipper/containers.hpp>
 #include <clipper/logging.hpp>
+#include <clipper/metrics.hpp>
 #include <clipper/util.hpp>
 
+#include <boost/circular_buffer.hpp>
 #include <boost/thread.hpp>
 
 namespace clipper {
 
 ModelContainer::ModelContainer(VersionedModelId model, int container_id,
                                InputType input_type)
-    : model_(model), container_id_(container_id), input_type_(input_type) {}
+    : model_(model),
+      container_id_(container_id),
+      input_type_(input_type),
+      avg_throughput_per_milli_(0),
+      throughput_buffer_(THROUGHPUT_BUFFER_CAPACITY) {}
+
+void ModelContainer::update_throughput(size_t batch_size,
+                                       long total_latency_micros) {
+  if (batch_size <= 0 || total_latency_micros <= 0) {
+    throw std::invalid_argument(
+        "Batch size and latency must be positive for throughput updates!");
+  }
+  boost::unique_lock<boost::shared_mutex> lock(throughput_mutex_);
+  double new_throughput = 1000 * (static_cast<double>(batch_size) /
+                                  static_cast<double>(total_latency_micros));
+  double old_total_throughput =
+      avg_throughput_per_milli_ * throughput_buffer_.size();
+  if (throughput_buffer_.size() == throughput_buffer_.capacity()) {
+    double oldest_throughput = throughput_buffer_.front();
+    double new_total_throughput =
+        (old_total_throughput - oldest_throughput + new_throughput);
+    avg_throughput_per_milli_ =
+        new_total_throughput / static_cast<double>(throughput_buffer_.size());
+  } else {
+    avg_throughput_per_milli_ =
+        (old_total_throughput + new_throughput) /
+        static_cast<double>(throughput_buffer_.size() + 1);
+  }
```

Comment:
Just to make sure I'm clear on what's happening here:

The else-branch will be taken when the circular buffer is first filling up. Once it's filled, the if-branch will be taken for the remainder of execution?

Can you add a brief comment inside the method explaining what's happening.